net: add set hostname function

MeisterBob opened the pull request:
This adds a function to set the hostname on runtime.

mt-gjoerges committed:
net: add set hostname function

This commit adds a function to set the hostname on runtime.

Signed-off-by: Gerhard Jörges <joerges@metratec.com>

jukkar reviewed (commented):
* We need tests for this new API in `tests/net/hostname` in a separate commit in this PR
* There are hostname users in zephyr that need to be able to start to use the new hostname. Could you check that they do not cache the hostname but use always the new one.

```
$ ack net_hostname_get
drivers/wifi/esp_at/esp.c
67:	snprintk(cmd, sizeof(cmd), "AT+CWHOSTNAME=\"%s\"", net_hostname_get());

lib/posix/uname.c
49:		strncpy(name->nodename, net_hostname_get(), sizeof(name->nodename));

subsys/mgmt/mcumgr/grp/os_mgmt/src/os_mgmt.c
639:			      (prior_output == true ? " %s" : "%s"), net_hostname_get());

subsys/net/ip/dhcpv4.c
164:	const char *hostname = net_hostname_get();

subsys/net/lib/dns/llmnr_responder.c
445:	const char *hostname = net_hostname_get();

subsys/net/lib/dns/mdns_responder.c
472:	const char *hostname = net_hostname_get();

subsys/net/lib/sockets/sockets_misc.c
13:	const char *p = net_hostname_get();

subsys/logging/backends/log_backend_net.c
115:		(void)strncpy(dev_hostname, net_hostname_get(), MAX_HOSTNAME_LEN);
```

rlubos commented on include/zephyr/net/hostname.h:
```diff
@@ -49,6 +53,18 @@ static inline const char *net_hostname_get(void)
+ *
+ * @return 0 if ok, <0 on error
+ */
+#if IS_ENABLED(CONFIG_NET_HOSTNAME_DYNAMIC)
```
Please use `#if defined()` with preprocessor directives. Plus, it'd be good to define a stub function (returning an error) in case the feature is not enabled.

rlubos commented on subsys/net/hostname.c:
```diff
@@ -34,9 +34,26 @@ static void trigger_net_event(void)
+	return _hostname;
 }
 
+#if IS_ENABLED(CONFIG_NET_HOSTNAME_DYNAMIC)
```
`#if defined()`

jukkar commented:
We have a separate `CONFIG_NET_HOSTNAME_UNIQUE_UPDATE` option that allows/denies update to hostname if the `CONFIG_NET_HOSTNAME_UNIQUE` is set.
It sounds like we should honor that and not allow hostname updates in that case. What people think about this? @rlubos

rlubos commented:
> We have a separate `CONFIG_NET_HOSTNAME_UNIQUE_UPDATE` option that allows/denies update to hostname if the `CONFIG_NET_HOSTNAME_UNIQUE` is set.
> It sounds like we should honor that and not allow hostname updates in that case. What people think about this? @rlubos

Yes, I think you're rigt, It'd likely make sense to make `CONFIG_NET_HOSTNAME_UNIQUE_UPDATE` and `CONFIG_NET_HOSTNAME_DYNAMIC` mutually exclusive.

mt-gjoerges committed:
tests: net: hostname: add test for net_hostname_set

Add test that sets the hostname.

Signed-off-by: Gerhard Jörges <joerges@metratec.com>

MeisterBob commented:
@jukkar  @rlubos  I addressed your issues and also added a test.

Regarding the `net_hostname_get` users: the only one that caches the hostname is the net logging backend.

rlubos commented on subsys/net/hostname.c:
```diff
@@ -17,14 +17,14 @@ LOG_MODULE_REGISTER(net_hostname, CONFIG_NET_HOSTNAME_LOG_LEVEL);
 #include <zephyr/net/net_mgmt.h>
 
-static char hostname[NET_HOSTNAME_SIZE];
+static char _hostname[NET_HOSTNAME_SIZE];
```
It was requested not to use underscore prefix for variable names.

MeisterBob replied:
What is the preferred naming style when I want to keep the `hostname` argument in `net_hostname_set`? Is `g_hostname` ok?

jukkar replied:
For example `char *host` would be fine as a function argument.

rlubos reviewed (commented):
One issue is still not addressed, looks good otherwise.

jukkar commented:
> Regarding the `net_hostname_get` users: the only one that caches the hostname is the net logging backend.

Shouldn't we fix it in this PR?

MeisterBob commented:
I would suggest using the hostname changed net_event, but I can't test it at the moment

https://github.com/metratec/zephyr/commit/0556c41f85df30eb83c1752d7928075355c6a9ee

jukkar commented:
> I would suggest using the hostname changed net_event, but I can't test it at the moment

That is certainly one option, but it requires that user has enabled net_mgmt support. Or if we select the net_mgmt support in logging, it causes this feature enabled even if no one else in the system is using it. That might be far fetched scenario but still possible. So I think the simplest and the smallest amount of system resources used is to call the logger function from hostname.c

mt-gjoerges committed:
logging: net: update hostname

This commit adds a function that updates the hostname displayed by the
net backend. It is called by the network stack when the hostname is
updated.

Signed-off-by: Gerhard Jörges <joerges@metratec.com>

MeisterBob commented:
I added log_backend_net_hostname_set().

jukkar commented on tests/net/hostname/src/main.c:
```diff
@@ -381,6 +381,13 @@ ZTEST(net_hostname, test_hostname_set)
+
+		ret = net_hostname_set("foobar", sizeof("foobar") - 1);
+		zassert_equal(ret, 0, "Could not set hostname (%d)", ret);
+	}
```
We should also verify that the set hostname is really set correctly, so need to call `net_hostname_get()` and verify that it is really the one we set earlier.
The getter could be called after this if() block so that it could be shared with the unique if that is easily doable.

jukkar reviewed (commented):
Thanks, looks good. Couple of minor nits.

jukkar reviewed (approved):
LGTM
